### PR TITLE
fix: rename typeParameters to typeArguments where needed

### LINKED ISSUE #146: Some properties named typeParameters instead of typeArguments
I've noticed several parts of the code referring to type arguments as type parameters. What's the reason for using `typeParameters` here? Seems like there might be some temporary stuff going on? Will this be renamed in the future?

```ts
if (node.typeArguments && node.typeArguments.length) {
    result.typeParameters = convertTypeArgumentsToTypeParameters(
        node.typeArguments
    );
}
```

https://github.com/typescript-eslint/typescript-eslint/blob/45d8ad3a9422365a19d29c1452a667aa2b8a1049/packages/typescript-estree/src/convert.ts#L2045-L2050

Some example code: [playground](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&showAST=es&code=M4ewtgpgYgrgdgYwC4EsRwDzCQJxXAcwD4AKASgG4ACAehqqQAsVgGBPABwioEMcCYkOEiotRiEDhwRkAGzZVpAMwhSIAEwYherHuy5UOfHpCSrxDRtx7YAsAChQkWIlToSARkq16TMXwEhETEESWk5BWVVaU0kbRteOF5+QQhhCyZrbCA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZYgQzMVoCNEAMwD20XoNhMy+SsKa0uYpujABtcNhyJo0UZAA0GzVk3ZInJnMIBbYbFQZz8AO6dCqQ6bNNO1xABMVSCZEADdtAyNTc2RCKQBBHWFnIM4XN1QoiABfKIBdDVzsoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

```ts
someFunction<string>(); // this type argument is incorrectly referred to as a type parameter in the ast
someFunction(1); // this argument is correctly referred to as an argument in the ast
```

Anyway, I just wanted to start a discussion about it so at least it's in the issue tracker. Also, FWIW, babel-eslint9 correctly refers to them as `typeArguments` when I'm looking at https://astexplorer.net, but I'm not so familiar with the babel or eslint ecosystem so I'm not sure how they're related (Edit: Looks like they call it `node.typeArguments.params` though...).

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
BREAKING CHANGE:
Adds duplicate properties on the AST and deprecates their old names

## PR Checklist

-   [x] Addresses an existing open issue: fixes #146
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Renames `typeParameters` to `typeArguments` in locations that are indeed providing type _arguments_.

To recap the terminology:

* An _argument_ is something you provide to a recipient, such as a type provided explicitly to a call expression
* A _parameter_ is how the recipient receives what you provide, such as a function declaration's generic type parameter

Marked as a breaking change even though the old, now-`@deprecated` property is still there.

### CODE REVIEW

**Comment by bradzacher** [comment]
Thinking about this in terms of the broader ecosystem and the impacts - I wonder if we want to give consumers one major version to migrate?
A hard cut over might be a bit difficult for the ecosystem to come handle as it would involve every single plugin updating their code (which could leave users in an impossible to upgrade state for some time).

How about this: this version we add the new property and we mark the old property as `@deprecated`, then in v7 we remove the old property entirely.

If we wanted we could also do something like define a getter that throws when in a test environment to help flag plugins that they should migrate.

WDYT?

**Comment by JoshuaKGoldberg** [comment]
Yeah that makes sense. The `@deprecated` idea is really nifty - seems like a good followup feature request.

File: packages/typescript-estree/tests/snapshots/typescript/expressions/tagged-template-expression-type-arguments.src.ts.shot [hunk 1]
```diff
@@ -80,7 +80,7 @@ Object {
           "type": "Identifier",
         },
         "type": "TaggedTemplateExpression",
-        "typeParameters": Object {
+        "typeArguments": Object {
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/snapshots/typescript/expressions/tagged-template-expression-type-arguments.src.ts.shot:83]: I'm confused as to why the unit tests are failing on this rename. https://github.com/typescript-eslint/typescript-eslint/actions/runs/4116598038/jobs/7106955244

```plaintext
typescript/expressions/tagged-template-expression-type-arguments.src.ts
    expect(received).toEqual(expected) // deep equality
    - Expected  - 1
    + Received  + 1
    @@ -75,11 +75,11 @@
                  3,
                ],
                "type": "Identifier",
              },
              "type": "TaggedTemplateExpression",
    -         "typeArguments": Object {
    +         "typeParameters": Object {
                "loc": Object {
                  "end": Object {
                    "column": 8,
                    "line": 1,
                  },

      90 |
      91 |     // Received = Babel, Expected = TSESTree
    > 92 |     expect(babelAst).toEqual(tsestreeAst);
         |                      ^
      93 |   });
      94 | });
```

@bradzacher am I missing something?

Comment (bradzacher) [reply packages/typescript-estree/tests/snapshots/typescript/expressions/tagged-template-expression-type-arguments.src.ts.shot:83]: Because these old tests essentially enforce that the AST are the same, rather than cataloguing the differences like the new tests.

Which is why they suck and I wanna get rid of them.

You should add a pre-process step to the parse here
https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/tests/ast-alignment/utils.ts

Comment (JoshuaKGoldberg) [reply packages/typescript-estree/tests/snapshots/typescript/expressions/tagged-template-expression-type-arguments.src.ts.shot:83]: ok that does it I'm scripting it

File: yarn.lock [hunk 2]
```diff
@@ -4343,6 +4343,61 @@
...
+    semver "^7.3.7"
+
+"@typescript-eslint/visitor-keys@5.52.0":
+  version "5.52.0"
+  resolved "https://registry.yarnpkg.com/@typescript-eslint/visitor-keys/-/visitor-keys-5.52.0.tgz#e38c971259f44f80cfe49d97dbffa38e3e75030f"
+  integrity sha512-qMwpw6SU5VHCPr99y274xhbm+PRViK/NATY6qzt+Et7+mThGuFSl/ompj2/hrBlRP/kq+BFdgagnOSgw9TB0eA==
+  dependencies:
+    "@typescript-eslint/types" "5.52.0"
+    eslint-visitor-keys "^3.3.0"
+
```
Comment (bradzacher) [inline yarn.lock:4400]: pichu_delet_this

this happens when there's a release we haven't merged into v6.
Yarn will attempt to install the old versions from npm.

I wonder if there's a way we can hard block yarn from doing this?

**Review by bradzacher — CHANGES_REQUESTED** [review]
> Yeah that makes sense. I'll put the @deprecated old properties in this PR (once I get the new ones working)

Once we've got this change, this is good-to-go I think.

File: yarn.lock (hunk 2 above)
Comment (armano2) [reply yarn.lock:4400]: wdyt about using https://yarnpkg.com/configuration/manifest/#resolutions on v6 branch?

Comment (bradzacher) [reply yarn.lock:4400]: I thought about it but it is kinda ugly cos you'd need to resolve things like
```
"@typescript-eslint/eslint-plugin": "file:./packages/eslint-plugin"
```

Which is what yarn workspaces are really supposed to do anyways

**Comment by JoshuaKGoldberg** [comment]
I think the lint & integration test failures are unrelated to this PR.
